### ISSUE TITLE
机器学习代码学习笔记，源码来自xiaotuidui，gpu训练实战

### ISSUE THREAD

**Issue by wwlachemer** [wwlachemer/blog#6]
import torch.optim   #数据优化器包
import torchvision   #计算机v   视觉处理包
from torch.nn import Sequential, Conv2d, MaxPool2d, Flatten, Linea   r#Sequential：这是一个容器，可以按顺序包含一系列的神经网络层， Conv2d卷积层用于指定输入通道，输出通道，卷积核，步幅，MaxPool2d最大池化，指定池化核的大小，步幅等，用于减小训练数据量，flatten，把输出数据变为一阶张量，Linear：全连接层
from torch.utils.data import DataLoader   #，用于tensorboard的tensor模块
from torch.utils.tensorboard import SummaryWriter：summarywriter    #，可以将用户的数据记录到日志当中，并且可以可视化查看
from torchvision import transforms#转换数据的包
# 引入神经网络
from torch import nn, optim
from tqdm import tqdm #进度条功能
# 准备训练和测试数据集
train_dataset = torchvision.datasets.CIFAR10(root="dataset",train=True,transform=transforms.ToTensor(),download=True)
#训练数据集保存在dataset，true表示加载CIFAR10，transform.totensor将数据plt图像或者numpy转化为张量，下载。
test_dataset = torchvision.datasets.CIFAR10(
    root="dataset",
    train=False,
    transform=transforms.ToTensor(),
    download=True)
# lenth 长度
train_data_len = len(train_dataset)#训练数据量
test_data_len = len(test_dataset)
print("训练数据集的长度：{len}".format(len=train_data_len))#string格式化输出
print("测试数据集的长度：{len}".format(len=test_data_len))
# 加载训练和测试数据集
train_dataloader = DataLoader(dataset=train_dataset,batch_size=64,shuffle=True)# 把数据放在dataloader里边，数据集是train_dataset ，batch_size,每一轮训练数据为64.shuffle是每一轮训练都要打乱。）
test_dataloader = DataLoader(dataset=test_dataset,batch_size=64,shuffle=True)

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")# torch.device:把数据上传gpu，如果gpu不可以用，那就转到cpu

print(device)#打印出设备
class Tudui(nn.Module):
#定义一个叫tudui的网络，继承nn.Module
    def __init__(self) -> None:#定义土堆初始化，
        super().__init__()#调用父类，让初始化继承nn.Module
        self.model = Sequential(
            Conv2d(3, 32, 5, stride=1, padding=2),
            MaxPool2d(2),
            Conv2d(32, 32, 5, stride=1, padding=2),
            MaxPool2d(2),
            Conv2d(32, 64, 5, stride=1, padding=2),
            MaxPool2d(2),
            Flatten(),
            Linear(1024, 64),
            Linear(64, 10)
        )#定义的model模型，使用了sequential容器，三个卷积层，三个池化层，一次平坦操作，两次全连接层，stride，卷积核移动步长，padding，数据外加两成像素

    def forward(self,x):
        x = self.model(x)
        return x
# 创建网络模型
tudui = Tudui()
tudui = tudui.to(device)
# 定义损失函数
loss_fn = nn.CrossEntropyLoss()#损失函数，交叉熵
loss_fn = loss_fn.to(device)
# 定义学习速率和优化器
learning_rate = 0.01#学习速率决定了模型在训练过程中的参数迭代的大小
optim = torch.optim.SGD(tudui.parameters(),learning_rate)
#torch.optim.SGD随机梯度下降，tutui.parameter获取全部参数。传递给SGD用于优化
# 设置训练网络的一些参数
# 记录训练的次数(累计每一批次，不是所有图片）
total_train_step = 0
# 记录测试的次数
total_test_step = 0
# 训练轮数
epoch = 15

writer = SummaryWriter("logs")
# 训练数据
for i in range(epoch):
    print("-----第{}轮训练开始-----".format(i+1))

    # 训练一轮数据
    tudui.train()
    for data in tqdm(train_dataloader):#遍历数据在train_dataloader中的数据，并且生成进度条
        imgs,targets = data#从当前批次数据中分别获取图像数据 imgs 和标签数据 targets
        imgs = imgs.to(device)
        targets = targets.to(device)
        output = tudui(imgs)#输出当前数据
        # 计算损失
        loss = loss_fn(output,targets)
        # 优化器调优
        optim.zero_grad()#清零优化器的梯度
        loss.backward()#损失函数反向传播
        optim.step()#优化器更新梯度
        total_train_step += 1
        # 避免输出太多
        if total_train_step % 100 ==0:#每100步开始下一步
            print("训练次数：{}，Loss：{}".format(total_train_step,loss.item()))
            writer.add_scalar("train_loss",loss.item(),total_train_step)#写入日志tensorboard

    # 测试步骤开始
    tudui.eval()#设置为评估模式
    total_test_loss = 0#初始化测试损失和测试值
    total_acc = 0
    with torch.no_grad(): #避免调优，不要设置梯度算法

        for data in test_dataloader:
            imgs, targets = data
            imgs = imgs.to(device)
            targets = targets.to(device)
            outputs = tudui(imgs)
            # 计算损失
            loss = loss_fn(outputs, targets)
            total_test_loss += loss
            # argmax为1表示，方向是横向比较
            acc = (outputs.argmax(1)==targets).sum()
            total_acc += acc#累加准确率，用于后续计算平均准确率

        print("整体测试集上Loss：{}".format(total_test_loss))
        print("整体测试集上正确率：{}".format(total_acc/test_data_len))

        writer.add_scalar("test_loss", total_test_loss, total_test_step)
        writer.add_scalar("test_acc", total_acc, total_test_step)
        total_test_step += 1
# 保存模型
torch.save(tudui,"tuidui.pth")
writer.close()
